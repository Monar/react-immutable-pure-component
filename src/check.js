import { is } from 'immutable';
import { getIn } from './utils';

export function check(config, prev = {}, next = {}, label) {
  let checkItem = createChecker(prev, next, label);
  let checklist = config || Object.keys({ ...next, ...prev });

  return checklist.every(checkItem);
}

function createChecker(prev, next) {
  return function(name) {
    if (typeof name === 'string') {
      return is(next[name], prev[name]);
    } else if (Array.isArray(name)) {
      return is(getIn(next, name), getIn(prev, name));
    }

    throw new TypeError('Invalid key: expected Array or string: ' + name);
  };
}
