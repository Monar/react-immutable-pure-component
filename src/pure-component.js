import React from 'react';
import { check } from './utils';

export class ImmutablePureComponent extends React.Component {
  shouldComponentUpdate(nextProps, nextState = {}) {
    return (
      !check(this.updateOnProps, this.props, nextProps, 'updateOnProps') ||
      !check(this.updateOnStates, this.state, nextState, 'updateOnStates')
    );
  }
}
