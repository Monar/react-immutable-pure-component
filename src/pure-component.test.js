/* eslint-env node, jest */
import { ImmutablePureComponent } from './pure-component';
import * as checkLib from './check';

let checkSpy = jest.spyOn(checkLib, 'check');

const getTestComponent = (props = {}, state) => {
  const Component = new ImmutablePureComponent();
  Component.props = props;
  Component.state = state;
  return Component;
};

describe('ImmutablePureComponent', () => {
  afterEach(() => {
    checkSpy.mockClear();
  });

  it('call check with everything is the same', () => {
    let props = { a: { b: 0 }, c: 10 };
    let state = { a: 'state', c: { d: 0 } };
    let Component = getTestComponent(props, state);
    Component.updateOnProps = ['c', ['a', 'b']];
    Component.updateOnStates = [['c', 'd'], 'a'];

    let result = Component.shouldComponentUpdate({ ...props }, { ...state });

    expect(checkSpy.mock.calls.length).toBe(2);
    expect(result).toBe(false);
    expect(checkSpy.mock.calls[0]).toMatchSnapshot('first call');
    expect(checkSpy.mock.calls[1]).toMatchSnapshot('second call');
  });

  it('call check with props different', () => {
    let props = { a: { b: 0 }, c: 10 };
    let state = { a: 'state', c: { d: 0 } };
    let Component = getTestComponent(props, state);
    Component.updateOnProps = ['c', ['a', 'b']];
    Component.updateOnStates = [['c', 'd'], 'a'];

    let result = Component.shouldComponentUpdate({ a: {} }, { ...state });

    expect(checkSpy.mock.calls.length).toBe(1);
    expect(result).toBe(true);
    expect(checkSpy.mock.calls[0]).toMatchSnapshot();
  });

  it('call check with state different', () => {
    let props = { a: { b: 0 }, c: 10 };
    let state = { a: 'state', c: { d: 0 } };
    let Component = getTestComponent(props, state);
    Component.updateOnProps = ['c', ['a', 'b']];
    Component.updateOnStates = [['c', 'd'], 'a'];

    let result = Component.shouldComponentUpdate(
      { ...props },
      { ...state, a: 'changed' },
    );

    expect(checkSpy.mock.calls.length).toBe(2);
    expect(result).toBe(true);
    expect(checkSpy.mock.calls[0]).toMatchSnapshot('first call');
    expect(checkSpy.mock.calls[1]).toMatchSnapshot('second call');
  });
});
