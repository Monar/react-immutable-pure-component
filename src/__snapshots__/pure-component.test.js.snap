// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImmutablePureComponent call check with everything is the same: first call 1`] = `
Array [
  Array [
    "c",
    Array [
      "a",
      "b",
    ],
  ],
  Object {
    "a": Object {
      "b": 0,
    },
    "c": 10,
  },
  Object {
    "a": Object {
      "b": 0,
    },
    "c": 10,
  },
  "updateOnProps",
]
`;

exports[`ImmutablePureComponent call check with everything is the same: second call 1`] = `
Array [
  Array [
    Array [
      "c",
      "d",
    ],
    "a",
  ],
  Object {
    "a": "state",
    "c": Object {
      "d": 0,
    },
  },
  Object {
    "a": "state",
    "c": Object {
      "d": 0,
    },
  },
  "updateOnStates",
]
`;

exports[`ImmutablePureComponent call check with props different 1`] = `
Array [
  Array [
    "c",
    Array [
      "a",
      "b",
    ],
  ],
  Object {
    "a": Object {
      "b": 0,
    },
    "c": 10,
  },
  Object {
    "a": Object {},
  },
  "updateOnProps",
]
`;

exports[`ImmutablePureComponent call check with state different: first call 1`] = `
Array [
  Array [
    "c",
    Array [
      "a",
      "b",
    ],
  ],
  Object {
    "a": Object {
      "b": 0,
    },
    "c": 10,
  },
  Object {
    "a": Object {
      "b": 0,
    },
    "c": 10,
  },
  "updateOnProps",
]
`;

exports[`ImmutablePureComponent call check with state different: second call 1`] = `
Array [
  Array [
    Array [
      "c",
      "d",
    ],
    "a",
  ],
  Object {
    "a": "state",
    "c": Object {
      "d": 0,
    },
  },
  Object {
    "a": "changed",
    "c": Object {
      "d": 0,
    },
  },
  "updateOnStates",
]
`;
